matrix-client-images.el

```diff
@@ -0,0 +1,96 @@
+                       ;; Disabling multibyte is required for reading binary data.
+                       ;; FIXME: require or autoload this function
+                       (mm-disable-multibyte)
+                       ;; Point is where the body starts, after the headers
+                       (buffer-substring (point) (point-max))))
+               ((eieio buffer) room))
+    (with-current-buffer buffer
+      ;; Rescale image in room buffer to get proper size
+      (matrix-client-rescale-image data))))
+
+(defun matrix-client-rescale-image (data)
+  "Rescale DATA, if too big, to fit the current buffer."
+  ;; Copied from image.el
+  (if (not (and (fboundp 'imagemagick-types)
+                (get-buffer-window (current-buffer))))
+      (create-image data nil t :ascent 100)
+    (let ((edges (window-inside-pixel-edges
+                  (get-buffer-window (current-buffer))))
+          (max-width (window-pixel-width))
+          (max-height (/ (window-pixel-height) 2)))
+      (create-image data 'imagemagick 'data-p
+                    :max-width max-width
+                    :max-height max-height))))
+
+(cl-defmethod matrix-client-insert-image-callback (&key (room matrix-client-room) message-id url
+                                                        data error-thrown symbol-status response
+                                                        &allow-other-keys)
+  "Insert image into proper place at URL in message MESSAGE-ID in ROOM.
+Image is passed from parser as DATA, which should be an image
+object made with `create-image'.  This function should be called
+as an async callback when the image is downloaded."
+  (with-slots (buffer) room
+    (with-current-buffer buffer
+      ;; Starting with last message, search backward to find message
+      (cl-loop initially do (goto-char (point-max))
+               for event_id = (get-text-property (point) 'event_id)
+               until (equal event_id message-id)
+               do (goto-char (previous-single-property-change (point) 'event_id)))
+      ;; Find beginning and end of message text
+      (let* ((beg (point))
+             (end (next-single-property-change beg 'event_id))
+             (inhibit-read-only t)
+             (props (text-properties-at beg))
+             (string (with-temp-buffer
+                       ;; NOTE: These space characters before the newlines
+                       ;; are REQUIRED for some reason.  Without them,
+                       ;; images disappear seemingly at random when
+                       ;; the user scrolls, moves point, or types.
+                       (insert " \n")
+                       (insert-image data)
+                       (insert " \n")
+                       (apply #'propertize (buffer-string) props))))
+        ;; Find URL and insert image after it
+        (goto-char beg)
+        (re-search-forward (regexp-quote url) end)
+        (goto-char (match-end 0))
+        (insert string)))))
```

Comment:
After we insert the image, maybe we should (goto (point-max)), when an image is inserted, point seems to move up one line before the image for me. Is it the same for you?

Reply:
Well, this runs asynchronously, so there may be later messages in the buffer when this runs.  It also runs within a save-excursion, I think.  A `(forward-char)` might be called for, though.  I'll look into this.  Thanks.

Reply:
Yep, that seems to have fixed it.